File: internal/tools/list_directory.go
package tools

import (
	"context"
	"fmt"
	"strings"

	"github.com/mark3labs/mcp-go/mcp"
	"github.com/mark3labs/mcp-go/server"
	"github.com/nanami7777777/gitlab-code-reader-mcp/internal/gitlab"
)

func ListDirectory(client *gitlab.Client) (mcp.Tool, server.ToolHandlerFunc) {
	tool := mcp.NewTool("gl_list_directory",
		mcp.WithDescription("List directory contents in a GitLab repository. Use depth=1 for quick browse, depth=2-3 for project structure overview."),
		mcp.WithString("project_id", mcp.Required(), mcp.Description("GitLab project ID or path")),
		mcp.WithString("path", mcp.Description("Directory path. Default: repository root")),
		mcp.WithString("ref", mcp.Description("Branch, tag, or commit SHA")),
		mcp.WithNumber("depth", mcp.Description("Recursion depth. Default: 1, max: 3")),
	)

	handler := func(ctx context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
		args := getArgs(req.Params.Arguments)
		projectID := argStr(args, "project_id")
		basePath := argStr(args, "path")

		ref := argStr(args, "ref")
		if ref == "" {
			r, err := client.DefaultBranch(projectID)
			if err != nil {
				return mcp.NewToolResultError(fmt.Sprintf("❌ Error: %v", err)), nil
			}
			ref = r
		}

		depth := 1
		if v, ok := argFloat(args, "depth"); ok && v > 0 {
			depth = minInt(int(v), 3)
		}

		recursive := depth > 1
		tree, err := client.GetTree(projectID, basePath, ref, recursive)
		if err != nil {
			return mcp.NewToolResultError(fmt.Sprintf("❌ Error: %v", err)), nil
		}

		// filter by depth
		baseDepth := 0
		if basePath != "" {
			baseDepth = strings.Count(basePath, "/") + 1
		}
		var filtered []gitlab.TreeItem
		for _, item := range tree {
			itemDepth := strings.Count(item.Path, "/") + 1 - baseDepth
			if itemDepth <= depth {
				filtered = append(filtered, item)
			}
		}

		maxItems := 200
		limited := filtered
		truncated := false
		if len(filtered) > maxItems {
			limited = filtered[:maxItems]
			truncated = true
		}

		displayPath := basePath
		if displayPath == "" {
			displayPath = "/"
		}
		header := fmt.Sprintf("Directory: %s (ref: %s, %d items)", displayPath, ref, len(filtered))

		var b strings.Builder
		b.WriteString(header + "\n" + strings.Repeat("─", 40) + "\n")
		for _, item := range limited {
			relPath := item.Path
			if basePath != "" && strings.HasPrefix(item.Path, basePath+"/") {
				relPath = item.Path[len(basePath)+1:]
			}
			indentLevel := strings.Count(relPath, "/")
			indent := strings.Repeat("  ", indentLevel)
			icon := "📄"
			if item.Type == "tree" {
				icon = "📁"
			}
			fmt.Fprintf(&b, "%s%s %s\n", indent, icon, item.Name)
		}

		if truncated {
			fmt.Fprintf(&b, "\n⚠️ Showing %d of %d items. Use a specific path to explore subdirectories.", maxItems, len(filtered))
		}
		return mcp.NewToolResultText(b.String()), nil
	}

	return tool, handler
}

File: src/tools/list-directory.ts
import { z } from "zod";
import type { GitLabClient } from "../gitlab/client.js";

export const listDirectorySchema = z.object({
  project_id: z.string().describe("GitLab project ID or path"),
  path: z.string().optional().describe("Directory path. Default: repository root"),
  ref: z.string().optional().describe("Branch, tag, or commit SHA"),
  depth: z.number().optional().describe("Recursion depth. Default: 1, max: 3"),
});

export type ListDirectoryInput = z.infer<typeof listDirectorySchema>;

export async function listDirectory(client: GitLabClient, input: ListDirectoryInput): Promise<string> {
  const ref = input.ref ?? (await client.getDefaultBranch(input.project_id));
  const depth = Math.min(input.depth ?? 1, 3);
  const basePath = input.path ?? "";

  const recursive = depth > 1;
  const tree = await client.getTree(input.project_id, basePath, ref, recursive);

  // Filter by depth
  const baseDepth = basePath ? basePath.split("/").length : 0;
  const filtered = tree.filter((item) => {
    const itemDepth = item.path.split("/").length - baseDepth;
    return itemDepth <= depth;
  });

  // Limit
  const maxItems = 200;
  const limited = filtered.slice(0, maxItems);
  const truncated = filtered.length > maxItems;

  // Build tree display
  const header = `Directory: ${basePath || "/"} (ref: ${ref}, ${filtered.length} items)`;

  const lines = limited.map((item) => {
    const relativePath = basePath && item.path.startsWith(basePath + "/")
      ? item.path.slice(basePath.length + 1)
      : item.path;
    const indent = "  ".repeat(relativePath.split("/").length - 1);
    const icon = item.type === "tree" ? "📁" : "📄";
    return `${indent}${icon} ${item.name}`;
  });

  let result = `${header}\n${"─".repeat(40)}\n${lines.join("\n")}`;

  if (truncated) {
    result += `\n\n⚠️ Showing ${maxItems} of ${filtered.length} items. Use a specific path to explore subdirectories.`;
  }

  return result;
}
